website/blog/2020-08-04-prestodb-and-hudi.md

```diff
@@ -0,0 +1,153 @@
+
+## Moving away from InputFormat.getSplits()
+
+While invoking `inputformat.getSplits()` was a simple integration, we noticed that this could cause a lot of RPC calls to namenode. There were several disadvantages to the previous approach. 
+
+1. The `InputSplit`s returned from Hudi are not enough. Presto requires to know the file status and block locations for each of the `InputSplit` returned. So this added 2 extra RPCs to the namenode for every split times the number of partitions loaded. Occasionally, backpressure can be observed if the namenode is under lots of pressure.
+2. Further, for every partition loaded (per `loadPartition()` call) in Presto split calculation, `HoodieParquetInputFormat.getSplits()` would be invoked. That caused redundant Hoodie table metadata listing, which otherwise can be reused for all partitions belonging to a table scanned from a query. 
+
+This led us to rethink the Presto-Hudi integration. At Uber, we changed this implementation by taking compile time dependency on Hudi and instantiated the `HoodieTableMetadata` once in the `BackgroundHiveSplitLoader` constructor. And leveraged Hudi Library APIs to filter the partition files instead of calling `HoodieParquetInputFormat.getSplits()`. This gave a significant reduction in the number of namenode calls in this path. 
+
+Towards generalizing this approach and making it available for the Presto-Hudi community, we added a new API in Presto’s `DirectoryLister` interface that would take in a `PathFilter` object. For Hudi tables, we supplied this PathFilter object - [HoodieROTablePathFilter](https://github.com/apache/incubator-hudi/blob/master/hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/HoodieROTablePathFilter.java), which would take care of filtering the files that Presto lists for querying Hudi tables and achieve the same results as Uber’s internal solution.
+
+This change is available since the 0.233 version of Presto and depends on the 0.5.1-incubating Hudi version. Since Hudi is brought into the compile time dependencies, we don't have to drop jars anymore into plugin dir like in the previous approach.
```

Comment:
Since Hudi is brought into the compile time dependencies, we don't have to drop jars anymore into plugin dir like in the previous approach. ->
Since Hudi is now a compile time dependency it is no longer necessary to provide Hudi jar files in the plugin directory.